File: node/lambda/src/data/repository/AniListRepository.ts
import { Media } from "../../domain/model/media";
import { MediaList } from "../../domain/model/mediaList";
import { MediaRepository } from "../../domain/repository/MediaRepository";

import { MEDIA_QUERY } from "../../graphql/queries/MediaQuery";
import { MEDIA_LIST_QUERY } from "../../graphql/queries/MediaListQuery";
import { MEDIA_LIST_MUTATION } from "../../graphql/queries/MediaListMutation";

import {
  ApolloClient,
  ApolloClientOptions,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client/core";
import { MediaListStatus as DataStatus } from "../../__generated__/graphql";
import { MediaListStatus } from "../../domain/model/mediaListStatus";

function domainToDataStatus(status: MediaListStatus): DataStatus {
  switch (status) {
    case "COMPLETED":
      return DataStatus.Completed;
    case "CURRENT":
      return DataStatus.Current;
    case "DROPPED":
      return DataStatus.Dropped;
    case "PAUSED":
      return DataStatus.Paused;
    case "PLANNING":
      return DataStatus.Planning;
    case "REPEATING":
      return DataStatus.Repeating;
  }
}

function getApolloClient(
  accessToken?: string,
): ApolloClient<NormalizedCacheObject> {
  const options: ApolloClientOptions<NormalizedCacheObject> = {
    cache: new InMemoryCache(),
    uri: "https://graphql.anilist.co",
  };
  if (accessToken !== undefined)
    options.headers = { Authorization: `Bearer ${accessToken}` };
  return new ApolloClient(options);
}

export const aniListRepository: MediaRepository = {
  searchMedia: async function (englishTitle: string): Promise<Media | null> {
    const client = getApolloClient();
    const response = await client.query({
      query: MEDIA_QUERY,
      variables: {
        mediaTitle: englishTitle,
      },
    });
    if (!response.data.Media) return null;
    return {
      id: response.data.Media.id,
      description: response.data.Media.description ?? "No description provided",
      episodes: response.data.Media.episodes ?? null,
      nextAiringEpisode: response.data.Media.nextAiringEpisode
        ? {
            episode: response.data.Media.nextAiringEpisode.episode,
            timeUntilAiring:
              response.data.Media.nextAiringEpisode.timeUntilAiring,
          }
        : null,
    };
  },
  searchMediaList: async function (
    accessToken: string,
    mediaId: number,
  ): Promise<MediaList | null> {
    const client = getApolloClient(accessToken);
    const response = await client.query({
      query: MEDIA_LIST_QUERY,
      variables: {
        mediaId: mediaId,
      },
    });
    if (!response.data.MediaList) return null;
    return {
      id: response.data.MediaList.id,
      mediaId: response.data.MediaList.mediaId!,
      progress: response.data.MediaList.progress!,
      score: response.data.MediaList.score!,
      status: response.data.MediaList.status!,
    };
  },
  updateMediaList: async function (
    accessToken: string,
    mediaId: number,
    variables: { progress?: number; score?: number; status?: MediaListStatus },
  ): Promise<void> {
    const client = getApolloClient(accessToken);
    await client.mutate({
      mutation: MEDIA_LIST_MUTATION,
      variables: {
        mediaId: mediaId,
        progress: variables.progress,
        score: variables.score,
        status:
          variables.status !== undefined
            ? domainToDataStatus(variables.status)
            : undefined,
      },
    });
  },
};

File: java/lambda/src/com/github/dannful/AniLexa/data/repository/AniListRepository.java
package com.github.dannful.AniLexa.data.repository;

import com.github.dannful.AniLexa.domain.model.Media;
import com.github.dannful.AniLexa.domain.model.MediaList;
import com.github.dannful.AniLexa.domain.model.MediaListStatus;
import com.github.dannful.AniLexa.domain.repository.AnimeRepository;
import com.github.dannful.AniLexa.utils.ClassSerializer;
import com.google.gson.Gson;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import org.json.JSONObject;

public class AniListRepository implements AnimeRepository {

  private final String accessToken;

  public AniListRepository(String accessToken) {
    this.accessToken = accessToken;
  }

  public AniListRepository() {
    this.accessToken = null;
  }

  private record KeyValuePair(String key, Object value) {}

  private String makeRequest(String body, KeyValuePair... variables) {
    HttpClient client = HttpClient.newBuilder().build();

    JSONObject bodyObject = new JSONObject();

    JSONObject variablesObject = new JSONObject();
    for (KeyValuePair keyValuePair : variables) variablesObject.put(
      keyValuePair.key,
      keyValuePair.value
    );
    bodyObject.put("variables", variablesObject);

    bodyObject.put("query", body);

    HttpRequest.Builder builder = HttpRequest.newBuilder(
      URI.create("https://graphql.anilist.co")
    )
      .header("Content-Type", "application/json")
      .header("Accept", "application/json")
      .POST(HttpRequest.BodyPublishers.ofString(bodyObject.toString()));

    if (this.accessToken != null) builder = builder.header(
      "Authorization",
      "Bearer " + this.accessToken
    );

    HttpRequest request = builder.build();

    try {
      String response = client
        .send(request, HttpResponse.BodyHandlers.ofString())
        .body();

      JSONObject object = new JSONObject(response);

      return object.get("data").toString();
    } catch (Exception e) {
      return null;
    }
  }

  private <T> void includeVariable(
    List<KeyValuePair> variables,
    String key,
    Optional<T> variable
  ) {
    variable.ifPresent(t -> variables.add(new KeyValuePair(key, t)));
  }

  @Override
  public Media searchMedia(String englishTitle) {
    String response = makeRequest(
      "query($mediaTitle: String) { Media(search: $mediaTitle, type: ANIME) { " +
      ClassSerializer.formatClassFields(Media.class) +
      " } }",
      new KeyValuePair("mediaTitle", englishTitle)
    );
    if (response == null) return null;
    Gson gson = new Gson();
    return gson.fromJson(
      new JSONObject(response).get("Media").toString(),
      Media.class
    );
  }

  @Override
  public MediaList searchMediaList(int mediaId) {
    String response = makeRequest(
      "query($mediaId: Int) { MediaList(mediaId: $mediaId) { " +
      ClassSerializer.formatClassFields(MediaList.class) +
      " } }",
      new KeyValuePair("mediaId", mediaId)
    );
    if (response == null) return null;
    Gson gson = new Gson();
    return gson.fromJson(
      new JSONObject(response).get("MediaList").toString(),
      MediaList.class
    );
  }

  @Override
  public void updateMediaList(
    int mediaId,
    Optional<Integer> progress,
    Optional<Float> score,
    Optional<MediaListStatus> status
  ) {
    List<KeyValuePair> variables = new ArrayList<>();
    variables.add(new KeyValuePair("mediaId", mediaId));
    includeVariable(variables, "progress", progress);
    includeVariable(variables, "score", score);
    includeVariable(variables, "status", status.map(Enum::name));
    makeRequest(
      "mutation($mediaId: Int, $progress: Int, $score: Float, $status: MediaListStatus) { SaveMediaListEntry(mediaId: $mediaId, progress: $progress, score: $score, status: $status) { progress } }",
      variables.toArray(new KeyValuePair[] {})
    );
  }
}
